ddclient

```diff
@@ -2298,57 +2281,72 @@ sub get_ip {
+## and no embedded leading zeros.
 ######################################################################
 sub is_ipv4 {
     my ($value) = @_;
     return (length($value // '') != 0) && ($value eq (ipv4_match($value) // ''));
 }
 
 ######################################################################
-## ipv4_match() extracts the first valid IPv4 address from given string
+## ipv4_match() extracts the first valid IPv4 address from given string.
+## Accepts leading zeros in the address but removes them in returned value
 ######################################################################
 sub ipv4_match {
-    (shift // '') =~ /\b((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\b/ai;
-    return $1;
+    (shift // '') =~ /\b((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3} ## first three bytes
+                         (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\b/xai;  ## last one
+    return ($1 // '') =~ s/\b0+\B//gr;                                     ## remove leading zeros
 }
 
 ######################################################################
-## is_ipv6() validates if string is valid IPv6 address
+## is_ipv6() validates if string is valid IPv6 address with no preceding
+## or trailing spaces/characters and no embedded leading zeros.
 ######################################################################
 sub is_ipv6 {
     my ($value) = @_;
-    # This little gem from http://home.deds.nl/~aeron/regex/
-    return $value =~ /^(((?=.*(::))(?!.*\3.+\3))\3?|([\dA-F]{1,4}(\3|:\b|$)|\2))(?4){5}((?4){2}|(((2[0-4]|1\d|[1-9])?\d|25[0-5])\.?\b){4})\z/ai;
+    return (length($value // '') != 0) && ($value eq (ipv6_match($value) // ''));
 }
 
 ######################################################################
-## ipv6_match determine ipv6 address from given string and return them
+## ipv6_match() extracts the first valid IPv6 address from given string
+## IPv6 must be in standard or compressed format.
+## Mixed IPv6/IPv4 not supported.
+## Accepts leading zeros in the address but removes them in returned value
 ######################################################################
 sub ipv6_match {
-    my $content = shift;
-    my $omits;
-    my $ip          = "";
-    my $linenumbers = 0;
-
-    my @values = split('\n', $content);
-    foreach my $val (@values) {
-        next unless $val =~ /((:{0,2}[A-F0-9]{1,4}){0,7}:{1,2}[A-F0-9]{1,4})/ai; # invalid char
-        my $parsed = $1;
-
-        # check for at least 7 colons
-        my $count_colon = () = $parsed =~ /:/g;
-        if ($count_colon != 7) {
-            # or one double colon
-            my $count_double_colon = () = $parsed =~ /::/g;
-            if ($count_double_colon != 1) {
-                next
-            }
-        }
-        return $parsed;
-    }
-    return;
+    (shift // '') =~ /(?<![:.\w])                                     ## Negative Look behind for word boundry
```

Comment:
Why not `\b`?

Reply:
Thats a good question.  I went with what the book said and didn't test alternatives (interestingly the book uses `\b` for IPv4 but not for IPv6).  I'll take another look.

Reply:
Okay, so what's actually going on here is that the regex is not looking for a word break, rather it is looking for (accepting) any characters that are valid within a IPv6 address... in this case `[:.\w]` (colon, period and a word character) now this is probably not exactly right for what we're doing -- I only test for standard and compressed formats, I do not accept mixed (IPv4 as the last 32 bits of an IPv6 address), so accepting a period is wrong (I suspect the book used this so that their regex examples were similar when accepting mixed format.  And `\w` also matches things like underscore which would be wrong also (the regex does not extract `::1` from `::1_` because the underscore is a word character).

So, for our purposes where we are looking only for a standard or compressed format IPv6 I think the lookahead / lookbehind should be for `[:0-9A-F]` case insensitive.  This is working for my list of test IPv6's.

Thanks for making me look at this.

Reply:
I've done further testing this morning and concluded that the book's lookahead/behind is actually correct.  If I omit the period then it matches `1111::123` from the string `1111::123.123.123.123` when it should reject the whole string (it is a mixed IPv6 rather than a standard or compressed IPv6).  Same for `[0-9A-F]` and `\w`.  It needs to be `\w` else the it matches `1111::123` from `1111::123_123_123_123` which again is wrong.  So I'm back to using what the book recommends.